File: msclust.r
library(kamila)
library("rjson")

k <- fromJSON(file = "k.json")$n_clusters

con_vars <- read.csv(file = "temp_continue.csv")
con_vars <- data.frame(scale(con_vars))

cat_vars_fac <- read.csv(file = "temp_cat.csv")

cat_vars_fac[] <- lapply(cat_vars_fac, factor)
cat_vars_dum <- dummyCodeFactorDf(cat_vars_fac)
cat_vars_dum <- data.frame(cat_vars_dum)


gms_res_hw <- gmsClust(con_vars, cat_vars_dum, nclust = k)

clusters <- gms_res_hw$results$cluster - 1

df <- cbind(
    con_vars,
    cat_vars_fac,
    cluster = clusters
)

write.csv(df, "temp_clustered.csv", row.names = FALSE)

File: algos/ModhaSpangler/msclust.py
import pandas as pd
import os
from subprocess import check_output

import json

def process(df, k):
    """Process Modha-Spangler Algorithm
    
    Interacts with Rscript, gmsclust function of kamila R Package"""

    # Write numerical and categorical to split CSV files, to fit kamila R package's needs
    numerical_columns = df.select_dtypes('number').columns
    categorical_columns = df.select_dtypes('object').columns
    df[categorical_columns].to_csv('temp_cat.csv',index=False)
    df[numerical_columns].to_csv('temp_continue.csv',index=False)

    # Pass the desired number of clusters to R through JSON
    json_data = { "n_clusters" : k}
    with open('k.json', 'w') as f:
        json.dump(json_data, f)

    # Process Data with R
    check_output("Rscript algos/ModhaSpangler/msclust.R", shell=True).decode()

    # Load clustered data
    df_out = pd.read_csv('temp_clustered.csv')

    df['cluster'] = df_out['cluster']

    # Remove temp files created to interact through R
    os.remove("temp_cat.csv")
    os.remove("temp_continue.csv")
    os.remove("temp_clustered.csv")
    os.remove("k.json")

    return df